hubcast/utils/git.py

```diff
@@ -1,17 +1,47 @@
-import os
 import subprocess
 
-# get configuration from environment.
-GIT_REPO_PATH = os.environ.get("HC_GIT_REPO_PATH")
 
+# https://git-scm.com/docs/gitcredentials#_custom_helpers
+def credential_helper(remote: str) -> str:
+    remote = remote.upper()
+    return f"""!f() {{ echo "username=${remote}_USERNAME"; echo "password=${remote}_TOKEN"; }};f"""
 
-def git(args):
-    """Executes a git command on the host system."""
-    result = subprocess.run(
-        ["git", "-C", f"{GIT_REPO_PATH}"] + args.split(),
-        stdout=subprocess.PIPE,
-        stderr=subprocess.STDOUT,
-        text=True,
-        check=True,
-    )
-    return result
+
+class Git:
+    def __init__(
+        self,
+        repo_path: str,
+        gh_username: str,
+        gh_token: str,
+        gl_username: str,
+        gl_token: str,
+        gl_instance_url: str,
+    ):
+        self.repo_path = repo_path
+        self.gl_url = gl_instance_url.removeprefix("https://")
+
+        self.gh_helper = credential_helper("github")
+        self.gl_helper = credential_helper("gitlab")
```

Comment:
We could definitely make them kwargs, any particular reason you'd prefer that over standard arguments?

At some level I think `git` has to be aware of GitHub and GitLab to correctly use the two different authentication schemes. Otherwise we can't impersonate a user pushing to GitLab and running under their HPC account.

Reply:
Sorry, I should have been more clear. I'll clarify here, but we can follow up with changes in another PR if you're amenable.

I meant to make the `Git` class initialization accept a set of credential helpers and `env` `dict`. For example:

```python
def credential_helper(remote: str, url: str) -> str:
    config_key = f"credential.{url}.helper"
    remote = remote.upper()
    return f"""{config_key}=!f() {{ echo "username=${remote}_USERNAME"; echo "password=${remote}_TOKEN"; }};f"""
    
class Git:
    def __init__(
        self,
        repo_path: str,
        env: dict,
        git_config: list,
    ):
    ...
    self.env = env
    self.config = "-c ".join(git_config)  # possibly a more elegant way to do this, just as an example
  
# elsewhere, outside
env = {
    "GITHUB_USERNAME": gh_username,
    "GITHUB_TOKEN": gh_token,
    "GITLAB_USERNAME": gl_username,
    "GITLAB_TOKEN": gl_token,
}

git_config = [
    credential_helper('github', 'github.com'),
    credential_helper('gitlab', 'gitlab.com'),
]

git = Git('/tmp', env, git_config=git_config)
```

Reply:
Done! Added you as a co-author to the commit.